### PR TITLE
Hunter improvements

### PR DESCRIPTION
**Pull request by Adrinalin4ik**

### CODE REVIEW

**Comment by Adrinalin4ik** [comment]
refered to #43

File: BlazorServer/run.bat [hunk 1]
```diff
@@ -1,6 +1,6 @@
 start "C:\Program Files (x86)\Google\Chrome\Application\chrome.exe" "http://localhost:5000"
-d:
-cd D:\GitHub\WowPixelBot\BlazorServer
+c:
+cd C:\Users\Alexey\Desktop\Projects\WowClassicGrindBot\BlazorServer
```
Comment (Xian55) [inline BlazorServer/run.bat:3]: revert this change it has no relation to the hunter class

File: .gitignore [hunk 2]
```diff
@@ -263,4 +263,6 @@ paket-files/
 
 # Python Tools for Visual Studio (PTVS)
 __pycache__/
-*.pyc
\ No newline at end of file
+*.pyc
+
+addons/WAB
```
Comment (Xian55) [inline .gitignore]: what's this `addons/WAB` and why has been added to this file?

File: Core/Goals/CombatGoal.cs [hunk 3]
```diff
@@ -39,7 +39,7 @@ public CombatGoal(ILogger logger, ConfigurableInput input, PlayerReader playerRe
             AddPrecondition(GoapKey.incombat, true);
             AddPrecondition(GoapKey.hastarget, true);
             AddPrecondition(GoapKey.targetisalive, true);
-            AddPrecondition(GoapKey.incombatrange, true);
+            //AddPrecondition(GoapKey.incombatrange, true);
```
Comment (Xian55) [inline Core/Goals/CombatGoal.cs:42]: The idea behind the `incombatrange` you can only use any spells if you are in range, otherwise you have to approach the target. Which is the `ApproachTargetGoal` responsibility.

File: Core/Goals/FollowRouteGoal.cs [hunk 4]
```diff
@@ -153,7 +154,7 @@ public override async Task PerformAction()
                 {
                     logger.LogError("wtf too far away kekw");
                     routeToWaypoint.Pop();
-                    return;
+                    // return;
```
Comment (Xian55) [inline Core/Goals/FollowRouteGoal.cs:157]: There's a hard limit in the `PathingAPI`, if the distance is higher then 100 yard it wont return any path.

File: Core/Goals/GoalFactory.cs [hunk 5]
```diff
@@ -85,6 +85,7 @@ public HashSet<GoapGoal> CreateGoals(ClassConfiguration classConfig, IBlacklist
                 if (classConfig.Loot)
                 {
                     var lootAction = new LootGoal(logger, input, addonReader.PlayerReader, addonReader.BagReader, stopMoving, classConfig, npcNameFinder);
+                   // var lootAction = new LootActions(logger, input, addonReader.PlayerReader, stopMoving, classConfig, castingHandler);
```
Comment (Xian55) [inline Core/Goals/GoalFactory.cs:88]: remove this line, the comment serves no meaningful info to others

File: Core/Goals/LootGoal.cs [hunk 6]
```diff
@@ -123,6 +123,12 @@ public override async Task PerformAction()
                     return;
                 }
                 */
+                if (!await Wait(4000, DiDEnteredCombat()))
+                {
+                    await AquireTarget();
+                    return;
+                }
+
```
Comment (Xian55) [inline Core/Goals/LootGoal.cs:131]: why would you want to wait 4 seconds to find a target here?

File: Core/Goals/LootGoal.cs [hunk 7]
```diff
@@ -227,6 +233,9 @@ private bool IsPlayerMoving(WowPoint lastPos)
 
         private async Task GoalExit()
         {
+            await input.TapLastTargetKey("Last target key");
+            await input.TapInteractKey("Interract key");
+            await Task.Delay(5000);
```
Comment (Xian55) [inline Core/Goals/LootGoal.cs:238]: The loot should not happen at this time frame. 

This function servers as the LootGoal as been finished either successfully or not and notify the other Goals the `GoapKey.shouldloot` has been changed to `false`

File: Core/Goals/PullTargetGoal.cs [hunk 8]
```diff
@@ -39,7 +39,8 @@ public PullTargetGoal(ILogger logger, ConfigurableInput input, PlayerReader play
             AddPrecondition(GoapKey.incombat, false);
             AddPrecondition(GoapKey.hastarget, true);
             AddPrecondition(GoapKey.pulled, false);
-            AddPrecondition(GoapKey.withinpullrange, true);
```
Comment (Xian55) [inline Core/Goals/PullTargetGoal.cs:42]: How can you be sure, you **can** pull the target if its not required to be in the `withinpullrange` ?

File: Core/Goals/PullTargetGoal.cs [hunk 9]
```diff
@@ -39,7 +39,8 @@ public PullTargetGoal(ILogger logger, ConfigurableInput input, PlayerReader play
...
+            AddPrecondition(GoapKey.targetisalive, true);
```
Comment (Xian55) [inline Core/Goals/PullTargetGoal.cs:42]: Why have to check if the target is alive? What's have to do with the this goal? 

To my best knowledge in `FollowRouteGal` searches for the possible target, but it does not acquire corpses

File: Core/NpcFinder/NpcNameFinder.cs [hunk 10]
```diff
@@ -120,9 +120,9 @@ public async Task FindAndClickNpc(int threshold, bool leftClick)
                 {
                     foreach (var location in locFindAndClickNpc)
                     {
-                        var clickPostion = bitmapProvider.DirectBitmap.ToScreenCoordinates(npc.ClickPoint.X + location.X, npc.ClickPoint.Y + location.Y);
+                        var clickPostion = bitmapProvider.DirectBitmap.ToScreenCoordinates(npc.ClickPoint.X + location.X, npc.ClickPoint.Y + location.Y + 30);
```
Comment (Xian55) [inline Core/NpcFinder/NpcNameFinder.cs:123]: please look what `locFindAndClickNpc` used for

File: Core/NpcFinder/NpcNameFinder.cs [hunk 11]
```diff
@@ -120,9 +120,9 @@ public async Task FindAndClickNpc(int threshold, bool leftClick)
...
                         mouseInput.SetCursorPosition(clickPostion);
-                        await Task.Delay(100);
+                        await Task.Delay(50);
```
Comment (Xian55) [inline Core/NpcFinder/NpcNameFinder.cs:125]: Are you 100% sure that scanning this quickly the Cursor will catches the desired `CursorClassification` ?

File: Core/Requirement/RequirementFactory.cs [hunk 12]
```diff
@@ -90,7 +90,10 @@ public Requirement GetRequirement(string name, string requirement)
             if (BuffDictionary.Count == 0)
             {
                 BuffDictionary = new Dictionary<string, Func<bool>>
-                {
+                {   
+                    // Target
+                    {  "hastarget", ()=> playerReader.HasTarget },
+                    {  "targetalive", ()=> !playerReader.PlayerBitValues.TargetIsDead },
```
Comment (Xian55) [inline Core/Requirement/RequirementFactory.cs:96]: Why do you expose these conditions if no class profile are using them? What point does they serve.

On top of the you have to update the readme.md in root, in order to other know about such changes

File: Json/class/Warlock_1.json [hunk 13]
```diff
@@ -85,4 +85,4 @@
       }
     ]
 }
-}
+}
```
Comment (Xian55) [inline Json/class/Warlock_1.json:88]: revert this change

File: Json/class/Warlock_1 copy.json [hunk 14]
```diff
@@ -0,0 +1,88 @@
+{
+  "ClassName": "Warlock",
+  "Loot": true,
+  "NPCMaxLevels_Below": 3,
+  "PathFilename": "test.json",
+  "PathThereAndBack": true,
+  "PathReduceSteps": false,
```
Comment (Xian55) [inline Json/class/Warlock_1 copy.json:7]: this class profile has no additional info compare to the `Warlock_1.json`. 
Please delete this.

File: PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs [hunk 15]
```diff
@@ -171,6 +171,10 @@ public static string[] GetArchiveNames(DataConfig dataConfig, Action<string> log
 
         public void SetContinent(string continent)
         {
+            if (continent == null) {
+                return;
+            }
+
```
Comment (Xian55) [inline PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs:177]: this one sounds like a workaround rather then an actual fix. By looking at the `Json/WorldToMap/WorldMapArea.json`

File: Json/WorldToMap/WorldMapArea.json [hunk 16]
```diff
@@ -59,6 +59,18 @@
...
+    "MapID": 0,
+    "AreaID": 36,
+    "AreaName": "Alterac",
+    "LocLeft": 783.3333,
+    "LocRight": -2016.6666,
+    "LocTop": 1500.0,
+    "LocBottom": -366.66666,
+    "UIMapId": 1415,
+    "Continent": "Azeroth"
+  },
```
Comment (Xian55) [inline Json/WorldToMap/WorldMapArea.json:73]: im seeing you have been added a duplicate for the `Alterac` with the only difference `"UIMapId": 1415,` changed. 

Question does `Alterac` has two maps `1415` and `1416` ?
or
The `1415` is not the correct one and supposed to be `1416`?

File: PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs [hunk 17]
```diff
@@ -213,9 +217,9 @@ public string SetZone(string zone)
             for (int i = 0; i < maps.recordCount; i++)
             {
                 int mapID = maps.GetInt(i, 0);
-                // logger.WriteLine("   ID:" + maps.GetInt(i, 0));
-                // logger.WriteLine(" File: " + maps.GetString(i, 1));
-                // logger.WriteLine(" Name: " + maps.GetString(i, 4)); // the file!!!
+                logger.WriteLine("   ID:" + maps.GetInt(i, 0));
+                logger.WriteLine(" File: " + maps.GetString(i, 1));
+                logger.WriteLine(" Name: " + maps.GetString(i, 4)); // the file!!!
```
Comment (Xian55) [inline PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs:222]: undo this change. it would rather just spam the output

File: .gitignore (hunk 2 above)
Comment (Xian55) [inline .gitignore:268]: why have to include this change? What's `WAB` ?

**Review by Xian55 — CHANGES_REQUESTED** [review]
Some of the key rules have been changed significantly which would have major effect on the other classes. Have you checked this changes on other classes? Beside warlock.

Some of the changes around the LootActions and LootActionGoal is not clear.

The `LootGoal` flow has been significantly changed compared to the previous form

File: Json/WorldToMap/WorldMapArea.json (hunk 16 above)
Comment (Adrinalin4ik) [reply Json/WorldToMap/WorldMapArea.json:73]: it has two, or even more.

File: PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs (hunk 15 above)
Comment (Adrinalin4ik) [reply PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs:177]: yes, but its too difficult to find info about locations, your json is not so good, alterac for example has 2 maps.

File: PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs (hunk 17 above)
Comment (Adrinalin4ik) [reply PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs:222]: the main reason why I did this to have a chance to see the location in order to fix it if it not found.

File: .gitignore (hunk 2 above)
Comment (Adrinalin4ik) [reply .gitignore:268]: Yes, I have renamed the addon I don't like the blizzard to catch me with this addon.

**Comment by Adrinalin4ik** [comment]
Actually, I'll not fix this stuff, I made a request to show people what I did to make a hunter works.

File: Core/NpcFinder/NpcNameFinder.cs (hunk 11 above)
Comment (Adrinalin4ik) [reply Core/NpcFinder/NpcNameFinder.cs:125]: No, it was just a test

File: Core/Goals/LootGoal.cs (hunk 7 above)
Comment (Adrinalin4ik) [reply Core/Goals/LootGoal.cs:238]: it's a crutch

File: Core/Goals/LootGoal.cs (hunk 6 above)
Comment (Adrinalin4ik) [reply Core/Goals/LootGoal.cs:131]: yes, It should be removed

File: .gitignore (hunk 2 above)
Comment (Xian55) [reply .gitignore:268]: I'm not completely sure what you trying to imply here but recently have been added a change where you have to customize the installed addon before its getting deployed into your wow installed addons folder, which means everyone has its own copy installed with a customized name(title), author, and even the in game setup command.

You are afraid of the Warden or the Blizzard client will scan your checked out repo location? This way of thinking the _anitcheat_ can scan any arbitrary folder...

File: Json/WorldToMap/WorldMapArea.json (hunk 16 above)
Comment (Xian55) [reply Json/WorldToMap/WorldMapArea.json:73]: Do you have any specific `UIMapId` in that regard beside the mentioned ones?

`UIMapId` of 1415 means Azeroth=Eastern Kingdoms. So in other zones suffering the same issue, then they are not gonna work at all since 1415 is now bound to `Alterac` zone bounds.

File: PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs (hunk 15 above)
Comment (Xian55) [reply PathingAPI/PPather/Triangles/MPQTriangleSupplier.cs:177]: All the data is coming from the following tables.

* data/uimap.csv
https://wow.tools/dbc/?dbc=uimap&build=2.5.1.38364&hotfixes=true#page=1

* data/worldmaparea.csv
https://wow.tools/dbc/?dbc=worldmaparea&build=2.4.3.8606#page=1

File: Json/WorldToMap/WorldMapArea.json (hunk 16 above)
Comment (Adrinalin4ik) [reply Json/WorldToMap/WorldMapArea.json:73]: I am not sure, but I guess that this JSON was exported from mpq from another client version.

Comment (Adrinalin4ik) [reply Json/WorldToMap/WorldMapArea.json:73]: I see that build is 2.4.3, but I used it for wow classic.

File: Core/Goals/PullTargetGoal.cs (hunk 8 above)
Comment (Adrinalin4ik) [reply Core/Goals/PullTargetGoal.cs:42]: Just use "I" btn and you'll be always in range.

File: BlazorServer/run.bat (hunk 1 above)
Comment (Adrinalin4ik) [reply BlazorServer/run.bat:3]: I think it should be in gitignore

File: Core/Goals/FollowRouteGoal.cs (hunk 4 above)
Comment (Adrinalin4ik) [reply Core/Goals/FollowRouteGoal.cs:157]: It causes crashes of the bot. Pathfinding sometimes works really bad. I faced with an issue where bot stayed about 5 yards away from the path point and the player just stayed or jumping without any movement. In log, I saw the message from this function. When I commented this line, the bot was able to get back into the path.

File: Core/Goals/CombatGoal.cs (hunk 3 above)
Comment (Adrinalin4ik) [reply Core/Goals/CombatGoal.cs:42]: My main idea to make bot to use approach time by time in the combat routine because incombatrange didn't work properly.

File: BlazorServer/run.bat (hunk 1 above)
Comment (Xian55) [reply BlazorServer/run.bat:3]: Yea completely right about this 👍

File: Json/WorldToMap/WorldMapArea.json (hunk 16 above)
Comment (Xian55) [reply Json/WorldToMap/WorldMapArea.json:73]: have you spotted any difference in the `1.13.7` vs `2.4.3` dataset from the `uimap` and `worldmaparea` when focusing `Alterac` ?
